agent/src/agent/modules/lookup.py

```diff
@@ -0,0 +1,82 @@
+from functools import wraps
+from agent.modules import data_source
+from typing import Callable, Optional, Any
+
+_sources = {}
+_cache = {}
+
+
+def provide(func):
+    """
+    Initializes lookup sources and cache and cleans them after the function execution
+    Wrapped function must have a pipeline argument of the type Pipeline in the first position
+    """
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        _clean()
+        _init_sources(args[0].config.get('lookup', {}))
+        res = func(*args, **kwargs)
+        # todo use finally or with
+        _clean()
+        return res
+
+    return wrapper
+
+
+def lookup(
+    lookup_name: str,
+    lookup_value: Any,
+    key_field: str,
+    value_field: str,
+    compare_function: Callable,
+) -> Optional[Any]:
+    """
+    Searches for a lookup_value in the key_field of a lookup using a provided compare function.
+    Returns a value from the value_field. Example:
+
+    aliases.csv
+    Alias     Type     Full_name
+    ifName_1  Network  Interface_name
+
+    res = lookup(
+        lookup_name='aliases.csv',
+        lookup_value='ifName',
+        key_field='Alias',
+        value_field='Full_name',
+        compare_function=startswith
+    )
+    res == 'Interface_name' // output: True
+    """
+    global _cache
+    if lookup_name not in _cache:
+        _cache[lookup_name] = _sources[lookup_name].get_data()
+    res = None
+    for obj in _cache[lookup_name]:
+        if compare_function(obj[key_field], lookup_value):
+            if res is not None:
+                raise Exception(
```

Comment:
How is it possible? Do we really need to fail all execution in this case?

Reply:
As an example it can be `startswith('port_')` and there could be two strings like `port_important` and `port_very_important`. I think in this case we cannot pick a random value from these two and there must always be only one value that matches provided criteria

Reply:
ok, but one more thing actually. Looping through every value is not efficient, maybe we also can store the result in a dict?